test(linter/plugins): debug build for tests

overlookmotel opened the pull request:
Generate a debug build, in addition to the normal release build.

In the debug build, `DEBUG` global is defined as `true` and `assertIsNonNull` does a runtime assertion. In release build, `DEBUG === false` and minifier removes debug-only code.

Run the tests in debug mode.

Debug build is generated in a separate directory from `dist`, so is *not* included in NPM package.

camc314 commented on apps/oxlint/vitest.config.ts:
```diff
@@ -2,6 +2,7 @@ import { configDefaults, defineConfig } from 'vitest/config';
   test: {
+    setupFiles: ['./vitest_setup.ts'],
     exclude: [...configDefaults.exclude, 'fixtures/**'],
   },
```
```suggestion
  test: {
    exclude: [...configDefaults.exclude, 'fixtures/**'],
  },
  define: {
    DEBUG: 'true'
  }
```

We can use vite's built in define plugin rather than a manual implementation

overlookmotel commented on apps/oxlint/tsdown.config.ts:
```diff
@@ -37,3 +38,13 @@ export default defineConfig([
+const debugConfigs = configs.map((config) => ({
+  ...config,
+  outDir: 'debug',
+  define: { DEBUG: 'true' },
```
Untrue. The test fixtures use the types from the debug build.

overlookmotel committed:
test(linter/plugins): debug build for tests (#15921)

Generate a debug build, in addition to the normal release build.

In the debug build, `DEBUG` global is defined as `true` and `assertIsNonNull` does a runtime assertion. In release build, `DEBUG === false` and minifier removes debug-only code.

Run the tests in debug mode.

Debug build is generated in a separate directory from `dist`, so is *not* included in NPM package.
